packages/web/src/store/application/ui/buy-audio/sagas.ts

```diff
@@ -371,87 +423,388 @@ Total: ${estimatedLamports.toNumber() / LAMPORTS_PER_SOL} SOL ($${
-    usd: estimatedUSD,
+    usd: estimatedUSD ?? '',
     sol: purchasedSOL,
     audio: purchasedAUDIO
   }
   const transactionDetails: TransactionDetails = {
     date: dayjs().format('MM/DD/YYYY'),
-    signature: transferTransactionId!,
+    signature: transferTransactionId,
     transactionType: TransactionType.PURCHASE,
-    method: TransactionMethod.COINBASE,
+    method:
+      PROVIDER_METHOD_MAP[localStorageState.provider ?? OnRampProvider.UNKNOWN],
     balance: postAUDIOBalance,
     change: purchasedAUDIO,
     metadata: transactionMetadata
   }
 
   yield* put(
     fetchTransactionDetailsSucceeded({
-      transactionId: transferTransactionId!,
+      transactionId: transferTransactionId,
       transactionDetails
     })
   )
   yield* call(saveUserBankTransactionMetadata, transactionMetadata)
+
+  // Clear local storage
+  yield* call(
+    [localStorage, localStorage.setJSONValue],
+    BUY_AUDIO_LOCAL_STORAGE_KEY,
+    {}
+  )
+}
+
+type PurchaseStepParams = {
+  estimatedSOL: AmountObject
+  connection: Connection
+  rootAccount: Keypair
+  provider: OnRampProvider
+  retryDelayMs?: number
+  maxRetryCount?: number
+}
+/**
+ * Executes the purchase step of the on-ramp
+ *
+ * @throws if cannot confirm the purchase
+ * @returns the new SOL balance for the rootAccount after the purchase succeeds
+ */
+function* purchaseStep({
+  estimatedSOL,
+  connection,
+  rootAccount,
+  provider,
+  retryDelayMs,
+  maxRetryCount
+}: PurchaseStepParams) {
+  // Cache current SOL balance
+  const initialBalance = yield* call(
+    [connection, connection.getBalance],
+    rootAccount.publicKey,
+    'finalized'
+  )
+
+  // Wait for on ramp finish
+  const result = yield* race({
+    success: take(onRampSucceeded),
+    canceled: take(onRampCanceled)
+  })
+
+  // If the user didn't complete the on ramp flow, return early
+  if (result.canceled) {
+    yield* put(make(Name.BUY_AUDIO_ON_RAMP_CANCELED, { provider }))
+    return
+  }
+  yield* put(make(Name.BUY_AUDIO_ON_RAMP_SUCCESS, { provider }))
+
+  // Wait for the SOL funds to come through
+  const newBalance = yield* call(pollForSolBalanceChange, {
+    rootAccount: rootAccount.publicKey,
+    initialBalance,
+    retryDelayMs,
+    maxRetryCount
+  })
+
+  // Get the purchase transaction
+  const signatures = yield* call(
+    [connection, connection.getSignaturesForAddress],
+    rootAccount.publicKey,
+    {
+      limit: 1
+    }
+  )
+  const purchaseTransactionId = signatures[0].signature
+
+  // Check that we got the requested SOL
+  const purchasedLamports = new BN(newBalance).sub(new BN(initialBalance))
+  if (purchasedLamports !== new BN(estimatedSOL.amount)) {
+    console.warn(
+      `Warning: Purchase SOL amount differs from expected. Actual: ${
+        new BN(newBalance).sub(new BN(initialBalance)).toNumber() /
+        LAMPORTS_PER_SOL
+      } SOL. Expected: ${estimatedSOL.uiAmountString} SOL.`
+    )
+  }
+
+  const localStorage = yield* getContext('localStorage')
```

Comment:
curious how come on line 500 we don't get the localStorage the same way as here. i see we do it this way too on line 668 for example

Reply:
I believe we do, just way at the top of the method (line 428)

Reply:
ah ok